src/utils_ovs.c

```diff
@@ -0,0 +1,1294 @@
+  ovs_callback_t *pre_cb = del_cb->prev;
+  ovs_callback_t *next_cb = del_cb->next;
+
+  pthread_mutex_lock(&pdb->mutex);
+  if (next_cb)
+    next_cb->prev = del_cb->prev;
+
+  if (pre_cb)
+    pre_cb->next = del_cb->next;
+  else
+    pdb->remote_cb = del_cb->next;
+
+  free(del_cb);
+  pthread_mutex_unlock(&pdb->mutex);
+}
+
+/* Remove all callbacks form OVS DB object */
+static void ovs_db_callback_remove_all(ovs_db_t *pdb) {
+  pthread_mutex_lock(&pdb->mutex);
+  for (ovs_callback_t *del_cb = pdb->remote_cb; pdb->remote_cb;
+       del_cb = pdb->remote_cb) {
+    pdb->remote_cb = pdb->remote_cb->next;
+    free(del_cb);
+  }
+  pdb->remote_cb = NULL;
+  pthread_mutex_unlock(&pdb->mutex);
+}
+
+/* Get/find callback in OVS DB object by UID. Returns pointer
+ * to requested callback otherwise NULL is returned */
+static ovs_callback_t *ovs_db_callback_get(ovs_db_t *pdb, uint64_t uid) {
+  pthread_mutex_lock(&pdb->mutex);
+  for (ovs_callback_t *cb = pdb->remote_cb; cb != NULL; cb = cb->next)
+    if (cb->uid == uid) {
+      pthread_mutex_unlock(&pdb->mutex);
+      return cb;
+    }
+  pthread_mutex_unlock(&pdb->mutex);
+  return NULL;
+}
+
+/* Send all requested data to the socket. Returns 0 if
+ * ALL request data has been sent otherwise negative value
+ * is returned */
+static int ovs_db_data_send(const ovs_db_t *pdb, const char *data, size_t len) {
+  ssize_t nbytes = 0;
+  size_t rem = len;
+  size_t off = 0;
+
+  while (rem > 0) {
+    if ((nbytes = send(pdb->conn.sock, data + off, rem, 0)) <= 0)
+      return (-1);
+    rem -= (size_t)nbytes;
+    off += (size_t)nbytes;
+  }
+  return (0);
+}
+
+/* Parse OVS server URL.
+ * Format of the URL:
+ *   "tcp:a.b.c.d:port" - define TCP connection (INET domain)
+ *   "unix:file" - define UNIX socket file (UNIX domain)
+ */
+static int ovs_db_url_parse(const char *surl, ovs_conn_t *conn) {
+  ovs_conn_t tmp_conn;
+  char *nexttok = NULL;
+  char *in_str = NULL;
+  char *saveptr;
+  int ret = 0;
+
+  /* sanity check */
+  if ((surl == NULL) || (strlen(surl) < 1))
+    return (-1);
+
+  /* parse domain */
+  tmp_conn = *conn;
+  in_str = sstrdup(surl);
+  if ((nexttok = strtok_r(in_str, ":", &saveptr)) != NULL) {
+    if (strcmp("tcp", nexttok) == 0) {
+      tmp_conn.domain = AF_INET;
+      tmp_conn.type = SOCK_STREAM;
+      tmp_conn.addr_size = sizeof(tmp_conn.addr.s_inet);
+    } else if (strcmp("unix", nexttok) == 0) {
+      tmp_conn.domain = AF_UNIX;
+      tmp_conn.type = SOCK_STREAM;
+      tmp_conn.addr_size = sizeof(tmp_conn.addr.s_unix);
+    } else
+      goto failure;
+  } else
+    goto failure;
+
+  /* parse url depending on domain */
+  if ((nexttok = strtok_r(NULL, ":", &saveptr)) != NULL) {
+    if (tmp_conn.domain == AF_UNIX) {
+      /* <UNIX-NAME> */
+      tmp_conn.addr.s_inet.sin_family = AF_UNIX;
+      sstrncpy(tmp_conn.addr.s_unix.sun_path, nexttok, strlen(nexttok) + 1);
+    } else {
+      /* <IP:PORT> */
```

Comment:
This doesn't handle IPv6.

I'd recommend to implement a config option that takes either one or two arguments. The first argument is the address, the second argument an optional port, for example:

```
Address "127.0.0.1" "6640"
Address "2001:DB8::c011:ec7d" "6640"
Address "2001:DB8::c011:ec7d" "service-name"
Address "unix:/path/to/socket"
```

You can then easily detect UNIX sockets: `strncmp (arg1, "unix:", strlen("unix:")) == 0`

Otherwise, call `getaddrinfo()` to get a `struct sockaddr`: `getaddrinfo(arg1, arg2, hints, &res)`